ext/opentelemetry-ext-pymemcache/src/opentelemetry/ext/pymemcache/__init__.py

```diff
@@ -0,0 +1,143 @@
+Usage
+-----
+.. code-block:: python
+    from opentelemetry import trace
+    from opentelemetry.sdk.trace import TracerProvider
+    from opentelemetry.ext.pymemcache import PymemcacheInstrumentor
+    trace.set_tracer_provider(TracerProvider())
+    from pymemcache.client.base import Client
+    client = Client(('localhost', 11211))
+    client.set('some_key', 'some_value')
+API
+---
+"""
+# pylint: disable=no-value-for-parameter
+
+import logging
+
+import pymemcache
+from wrapt import ObjectProxy
+from wrapt import wrap_function_wrapper as _wrap
+
+from opentelemetry.auto_instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.ext.pymemcache.util import (
+    _get_address_attributes,
+    _get_query_string,
+)
+from opentelemetry.ext.pymemcache.version import __version__
+from opentelemetry.trace import SpanKind, get_tracer
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+logger = logging.getLogger(__name__)
+
+_DEFAULT_SERVICE = "memcached"
+_RAWCMD = "db.statement"
+_CMD = "memcached.command"
+COMMANDS = [
+    "set",
+    "set_many",
+    "add",
+    "replace",
+    "append",
+    "prepend",
+    "cas",
+    "get",
+    "get_many",
+    "gets",
+    "gets_many",
+    "delete",
+    "delete_many",
+    "incr",
+    "decr",
+    "touch",
+    "stats",
+    "version",
+    "flush_all",
+    "quit",
+    "set_multi",
+    "get_multi",
+]
+
+
+def _set_connection_attributes(span, instance):
+    for key, value in _get_address_attributes(instance).items():
+        span.set_attribute(key, value)
+
+
+def _with_tracer_wrapper(func):
+    """Helper for providing tracer for wrapper functions.
+    """
+
+    def _with_tracer(tracer, cmd):
+        def wrapper(wrapped, instance, args, kwargs):
+            # prevent double wrapping
+            if hasattr(wrapped, "__wrapped__"):
+                return wrapped(*args, **kwargs)
+
+            return func(tracer, cmd, wrapped, instance, args, kwargs)
+
+        return wrapper
+
+    return _with_tracer
+
+
+@_with_tracer_wrapper
+def _wrap_cmd(tracer, cmd, wrapped, instance, args, kwargs):
+    with tracer.start_as_current_span(
+        _CMD, kind=SpanKind.INTERNAL, attributes={}
+    ) as span:
+        try:
+            vals = _get_query_string(args)
+            query = "{}{}{}".format(cmd, " " if vals else "", vals)
+            span.set_attribute(_RAWCMD, query)
+
+            _set_connection_attributes(span, instance)
+        except Exception:  # pylint: disable=broad-except
+            pass
+
+        return wrapped(*args, **kwargs)
+
+
+def _unwrap(obj, attr):
+    func = getattr(obj, attr, None)
+    if func and isinstance(func, ObjectProxy) and hasattr(func, "__wrapped__"):
+        setattr(obj, attr, func.__wrapped__)
+
+
+class PymemcacheInstrumentor(BaseInstrumentor):
+    """An instrumentor for pymemcache
+    See `BaseInstrumentor`
+    """
+
+    def _instrument(self, **kwargs):
```

Comment:
I think that sounds possible, I think some of the dd-trace-py instrumentations expose something to that effect, and it could have some value, but I don't think it's trivial to add. Is that convention established anywhere in opentelemetry-python at the moment?

Reply:
good question, I don't think it's written down anywhere, I just see it a lot (all db instrumentations, all webserver instrumentations). There's probably no need to add it in this PR unless there was actually a use case where you would want to instrument one client but not others

Reply:
no strong convention there, but see flask for an example:https://github.com/open-telemetry/opentelemetry-python/blob/master/ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py#L194

One thing is that "app" is probably a misnomer in the context of memcache, so the convention is loose at best.

Reply:
There is no convention for any other method outside `_instrument` and `_uninstrument`. Since these instrumentations deal with different third party libraries it is ok to add more methods as needed (like it was done for the Flask instrumentation) to better adapt the instrumentation to the characteristics of the third party libraries.